frontend/simple/js/state.js

```diff
@@ -65,29 +83,52 @@ const mutations = {
     state.position = position
   },
   pending (state, contractId) {
-    let index = state.pending.indexOf(contractId)
-    index === -1 && state.pending.push(contractId)
+    if (!state.contracts[contractId]) {
+      let index = state.pending.indexOf(contractId)
+      index === -1 && state.pending.push(contractId)
+    }
   }
 }
-
 // https://vuex.vuejs.org/en/getters.html
 // https://vuex.vuejs.org/en/modules.html
 const getters = {
   currentGroup (state) {
     return state[state.currentGroupId]
+  },
+  mailbox (state) {
+    let mailboxContract = getMailboxContract(state)
+    return mailboxContract && mailboxContract.messages
+  },
+  unreadMessageCount (state) {
+    let mailboxContract = getMailboxContract(state)
+    return mailboxContract && mailboxContract.messages.filter((msg) => !msg.read).length
   }
 }
 
 const actions = {
+  // Used to update contracts to the current state that the server is aware of
+  async syncContractWithServer ({dispatch}: {dispatch: Function}, contractId: string) {
+    let latest = await backend.latestHash(contractId)
+    let recent = await db.recentHash(contractId)
+    if (latest !== recent) {
+      let events = await backend.eventsSince(contractId, recent || contractId)
+      for (let i = 0; i < events.length; i++) {
+        let event = events[i]
+        event.contractId = contractId
+        await dispatch('handleEvent', event)
```

Comment:
I think, but haven't checked, that this loop is adding the event at `events[0]` twice. i.e. `eventsSince(recent)` is *inclusive*, not exclusive, so the event at `recent` is returned, but we already have it, so we should skip it.